### PR TITLE
feat(griffin): haversine distance function

### PR DESCRIPTION
**Pull request by jaugsburger**

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        mapValue.putDouble(this.valueIndex + 2, 0);
+    }
+
+    @Override
+    public void computeNext(MapValue mapValue, Record record) {
+        double lat1 = toRad(mapValue.getDouble(valueIndex));
+        double lon1 = toRad(mapValue.getDouble(valueIndex + 1));
+        double lat2 = toRad(this.latDegree.getDouble(record));
+        double lon2 = toRad(this.lonDegree.getDouble(record));
+        if (Double.isNaN(lat1) || Double.isNaN(lon1)) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java:65]: Function should be optimised for happy path, e.g. none are null. There will be two branch misprediction most of the time in current impl

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        mapValue.putDouble(this.valueIndex + 2, 0);
+    }
+
+    @Override
+    public void computeNext(MapValue mapValue, Record record) {
+        double lat1 = toRad(mapValue.getDouble(valueIndex));
+        double lon1 = toRad(mapValue.getDouble(valueIndex + 1));
+        double lat2 = toRad(this.latDegree.getDouble(record));
+        double lon2 = toRad(this.lonDegree.getDouble(record));
+        mapValue.putDouble(this.valueIndex, this.latDegree.getDouble(record));
```
[the remark this answers is not in the document]
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java]: done

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        mapValue.putDouble(this.valueIndex + 1, this.lonDegree.getDouble(record));
+        if (Double.isNaN(lat1) || Double.isNaN(lat2) || Double.isNaN(lon1) || Double.isNaN(lon2)) {
```
[the remark this answers is not in the document]
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java]: done

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            return;
+        }
+        double halfLatDist = (lat2 - lat1) / 2;
+        double halfLonDist = (lon2 - lon1) / 2;
+        double a = sin(halfLatDist) * sin(halfLatDist) + cos(lat1) * cos(lat2) * sin(halfLonDist) * sin(halfLonDist);
+        double c = 2 * atan2(sqrt(a), sqrt(1 - a));
+        double storedDistance = mapValue.getDouble(this.valueIndex + 2);
+        double distance = Double.isNaN(storedDistance) ? 0 : storedDistance;
```
[the remark this answers is not in the document]
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java]: removed check

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        columnTypes.add(ColumnType.DOUBLE);
+        columnTypes.add(ColumnType.DOUBLE);
+        columnTypes.add(ColumnType.DOUBLE);
+    }
+
+    @Override
+    public void setDouble(MapValue mapValue, double value) {
+        mapValue.putDouble(this.valueIndex, value);
+        mapValue.putDouble(this.valueIndex + 1, value);
+        mapValue.putDouble(this.valueIndex + 2, 0.0);
```
[the remark this answers is not in the document]
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java]: this is strange, this method is not being called in the tests, I'm looking into it

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunctionFactory.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.sql.Function;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.std.ObjList;
+
+public class HaversineDistDegreeGroupByFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "haversine_dist_degree(DD)";
```
[the remark this answers is not in the document]
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunctionFactory.java]: done

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java (hunk 1 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java]: done

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 7]
```diff
@@ -435,20 +498,51 @@ protected RecordCursor initFunctionsAndCursor(
         return cursor;
     }
 
+    private void interpolateBoundaryRange(long x1, long x2, Record record) {
+        //interpolating boundary
+        for (int i = 0, m = groupByDistanceFunctions.size(); i < m; i++) {
+            GroupByFunction function = groupByDistanceFunctions.getQuick(i);
+            MapValue value1 = findDataMapValue2(record, x1);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:505]: please add comment here

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java [hunk 8]
```diff
@@ -52,164 +53,148 @@ public HaversineDistDegreeGroupByFunction(int position, @NotNull Function latDeg
...
+        //set null to last item
+        saveLastItem(mapValue, Double.NaN, Double.NaN, Numbers.LONG_NaN);
+        //
+        saveDistance(mapValue, 0.0);
     }
 
-    private double getHaversineDistanceFromDegrees(double lat1Degrees, double lon1Degrees, double lat2Degrees, double lon2Degrees, double currentTotalDistance) {
+    private double calculateHaversineDistance(MapValue value1, MapValue value2) {
+        //value1
+        double lat1Degrees = getLastLatitute(value1);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java:179]: "Latitude"

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java [hunk 9]
```diff
@@ -52,164 +53,148 @@ public HaversineDistDegreeGroupByFunction(int position, @NotNull Function latDeg
...
 
-    @Override
-    public Function getLeft() {
-        return this.latDegree;
-    }
-
     @Override
     public void setDouble(MapValue mapValue, double value) {
-        mapValue.putDouble(this.valueIndex + 6, value);
+        saveDistance(mapValue, value);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java:164]: under which circumstances would this be called?

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 10]
```diff
@@ -363,6 +383,7 @@ public RecordCursor getCursor(SqlExecutionContext executionContext) {
                                                     // group-by function
                                                     // use current 'value' for record
                                                     computeYPoints(mapRecord, x1, x2value);
+                                                    interpolateDistanceRange(x1, x2, loSample, x1, mapRecord);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:386]: we should come up with nomenclature for these functions and use it consistently

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 11]
```diff
@@ -500,24 +522,24 @@ protected RecordCursor initFunctionsAndCursor(
...
-            for (int i = 0, m = groupByDistanceFunctions.size(); i < m; i++) {
-                GroupByFunction function = groupByDistanceFunctions.getQuick(i);
-                MapValue value1 = findDataMapValue2(record, x1);
-                MapValue value2 = findDataMapValue3(record, x2);
-                interpolatorGapDistanceFunctions.getQuick(i).interpolateGapAndStore(function, xValue, sampler.getBucketSize(), value1, value2);
+            final MapValue result = findDataMapValue(record, x);
+            assert result != null && result.getByte(0) == 1;
+            for (int i = 0; i < groupByTwoPointFunctionCount; i++) {
+                GroupByFunction function = groupByTwoPointFunctions.getQuick(i);
+                MapValue startValue = findDataMapValue2(record, x1);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:540]: This is excessive hash map scan. In some cases map values are already known, we shouldn't be looking up things that we don't have to

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java]: fixed

File: core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java (hunk 8 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/functions/groupby/HaversineDistDegreeGroupByFunction.java:179]: done

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java (hunk 11 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:540]: done

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 12]
```diff
@@ -142,13 +164,47 @@ public SampleByInterpolateRecordCursorFactory(
...
+                    throw SqlException.$(function.getPosition(), "Unsupported type: ").put(ColumnType.nameOf(function.getType()));
+            }
+        }
+
+        this.groupByTwoPointFunctionCount = groupByTwoPointFunctions.size();
+        for (int i = 0; i < groupByTwoPointFunctionCount; i++) {
+            GroupByFunction function = groupByTwoPointFunctions.getQuick(i);
+            switch (function.getType()) {
+                case ColumnType.BYTE:
+                    interpolatorGapTwoPointFunctions.add(InterpolationUtil.INTERPOLATE_GAP_BYTE);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:177]: The "two point" function interpolation API is the same for every result type.

What is the point of having collections of function references here?

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java (hunk 12 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java]: fixed

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 13]
```diff
@@ -525,25 +490,14 @@ protected RecordCursor initFunctionsAndCursor(
...
-
     private void interpolate(long lo, long hi, Record mapRecord, long x1, long x2, MapValue x1Value, MapValue x2value) {
         computeYPoints(x1Value, x2value);
         for (long x = lo; x < hi; x = sampler.nextTimestamp(x)) {
             final MapValue result = findDataMapValue3(mapRecord, x);
             assert result != null && result.getByte(0) == 1;
             for (int i = 0; i < groupByTwoPointFunctionCount; i++) {
                 GroupByFunction function = groupByTwoPointFunctions.getQuick(i);
-                interpolatorGapTwoPointFunctions.getQuick(i).interpolateGapAndStore(function, result, sampler.getBucketSize(), x1Value, x2value);
+                InterpolationUtil.INTERPOLATE_GAP.interpolateGapAndStore(function, result, sampler.getBucketSize(), x1Value, x2value);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:500]: inline

File: core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java [hunk 14]
```diff
@@ -553,6 +507,17 @@ private void interpolate(long lo, long hi, Record mapRecord, long x1, long x2, M
         }
     }
 
+    private void interpolateBoundaryRange(long x1, long x2, Record record) {
+        //interpolating boundary
+        for (int i = 0; i < groupByTwoPointFunctionCount; i++) {
+            GroupByFunction function = groupByTwoPointFunctions.getQuick(i);
+            MapValue startValue = findDataMapValue2(record, x1);
+            MapValue endValue = findDataMapValue3(record, x2);
+            InterpolationUtil.INTERPOLATE_BOUNDARY.interpolateBoundaryAndStore(function, sampler.nextTimestamp(x1), startValue, endValue, true);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java:516]: inline

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**